Add instrumentation and profiling API

illuhad opened the pull request:
This PR is based on PR #428, but introduces a redesign of the instrumentation and API, which allows for some simplifications. In particular, in this PR instrumentations can also be added by the background worker thread (e.g. the scheduler). This is done by collecting instrumentations until the scheduler has submitted the node.

Additionally, this PR moves to a design that relies more on existing idioms and functionality of the hipSYCL runtime.

All backends except for Level Zero are currently supported.

@fknorr feedback appreciated :)

TODO: Think about how we can access instrumentations of requirements and how we can handle multi-operation nodes in the future

fknorr committed:
Implement queue profiling

illuhad committed:
Merge branch 'rt-profiling' of https://github.com/fknorr/hipSYCL into develop

illuhad committed:
Redesign instrumentation API

illuhad committed:
Remove make_hip/cuda_event

illuhad committed:
Fix formatting

fknorr commented on include/hipSYCL/runtime/instrumentation.hpp:
```diff
@@ -0,0 +1,193 @@
+    return get_time_point().time_since_epoch().count();
+  }
+
+  double get_time_seconds() const {
```
`get_time_seconds` is never used. For best type-safety I suggest we remove both `get_ns_ticks` and `get_time_seconds`, doing the `profiler_clock::time_point -> uint64_t` conversion only at the top-level `event::get_profiling_info`

illuhad replied:
The idea was that we might need those conversions for future functionality that I want to work on, but maybe it makes more sense to have those conversions as free functions (which would also remove code duplication).

fknorr commented on include/hipSYCL/runtime/instrumentation.hpp:
```diff
@@ -0,0 +1,193 @@
+      _instrs;
+
+  mutable signal_channel _registration_complete_signal;
+  mutable std::atomic<bool> _is_registration_complete = false;
```
This flag looks redundant, we should be able to just call `_registration_complete_signal->has_signalled()`

illuhad replied:
The idea here is to cache whether the registration is complete (a similar pattern exists in the dag node implementation), but I guess it could be argued that this is a case of premature optimization.

fknorr replied:
I would expect futures to be implemented with atomics where possible anyway.

fknorr commented on src/runtime/omp/omp_instrumentation.cpp:
```diff
@@ -0,0 +1,62 @@
+}
+
+void omp_execution_start_timestamp::record_time() {
+  _time = profiler_clock::now();
```
I've been thinking some time about whether the access to `_time` from `record_time()` and `get_time_point()` are thread-safe or not. I would expect `signal::signal()` to synchronize-with `signal::wait()` for practical purposes, but cppref does not say anything about such a relationship between `std::promise::set_value()` and `std::shared_future::wait()`. As long as nobody calls `get_time_point()` before wait, it *should* be safe.

Since this is a brittle interface as-is, it might be best to make `signal_channel` generic and transport a value (here, a `profiler_clock::time_point`) to avoid this question altogether.

illuhad replied:
Do I understand correctly that you say that we cannot actually rely on `signal_channel::wait()` to synchronize with `signal_channel::signal()`? The whole point of the `signal_channel` class was to allow for such patterns, if it doesn't do that then I think we should rather fix `signal_channel`. 
Especially since it's not a new class: It has existed before and is e.g. used to implement events in the OpenMP backend (this PR only moves it to a separate header file to make it usable by the instrumentation API). So, if it's broken existing code relying on it will be broken as well.

fknorr replied:
I dug into the standard and *[futures.state]* does say that `shared_future::wait()` synchronizes-with `promise::set_value()`, so the use in `omp_node_event` is definitely file. The issue that remains IMO is that `omp_execution_*_timestamp` can be easily mis-used: Calling `record_time()` twice or calling `get_time_point()` before `wait()` will usually cause a data race because `_time` is not synchronized otherwise. The straight-forward solution should be to transport a value through `signal_channel` instead of the implicit `bool` it uses now.

illuhad replied:
Thanks for checking with the standard!

So the issue is that it could be misused. Calling `get_time_point()` before `wait()` cannot really happen because the `instrumentation_set` is deliberately designed to not grant direct access to instrumentations, and internally executing `wait()` before accessing results. This simplifies the implementation of the instrumentations, because such synchronization issues are now handled at a central place.
In this design, `instrumentation` objects should mainly be seen as a simple vehicle to store a result.

I have now made `record_time()` private such that it can only be accessed by the OpenMP instrumentation setup logic. It is after all more of an implementation detail of OpenMP instrumentation setup. There are now also asserts to check that it was not invoked before.

fknorr reviewed (commented):
Thank you for taking the time to look into profiling again! Overall, I like this approach. I was not thinking about potential deadlocks around waiting for instrumentation results.

illuhad committed:
Improve compatibility with earlier gcc versions

illuhad committed:
Move mechanism to get ns_ticks and seconds to profiler_clock

illuhad committed:
Remove redundant atomic flag to check if instrumentation setup phase is complete

illuhad committed:
Store executed operation in dag node

illuhad committed:
Properly implement queue profiling aspect and device info query

illuhad committed:
Consider omp_*_timestamp implementation detail of omp_instrumentation_setup

fknorr commented on include/hipSYCL/runtime/cuda/cuda_instrumentation.hpp:
```diff
@@ -0,0 +1,60 @@
+class cuda_event_time_delta {
+public:
+  profiler_clock::duration operator()(std::shared_ptr<dag_node_event> t0,
+                                      std::shared_ptr<dag_node_event> t1) const;
```
These parameters don't need to transfer ownership, `const dag_node_event &` should be enough

fknorr commented on include/hipSYCL/runtime/hip/hip_instrumentation.hpp:
```diff
@@ -0,0 +1,60 @@
+class hip_event_time_delta {
+public:
+  profiler_clock::duration operator()(std::shared_ptr<dag_node_event> t0,
+                                      std::shared_ptr<dag_node_event> t1) const;
```
`const dag_node_event &` should be enough here

fknorr commented on src/runtime/cuda/cuda_instrumentation.cpp:
```diff
@@ -0,0 +1,65 @@
+
+profiler_clock::duration
+cuda_event_time_delta::operator()(std::shared_ptr<dag_node_event> t0,
+                                  std::shared_ptr<dag_node_event> t1) const {
```
`const dag_node_event &` is sufficient

fknorr commented on src/runtime/hip/hip_instrumentation.cpp:
```diff
@@ -0,0 +1,64 @@
+namespace rt {
+
+profiler_clock::duration
+hip_event_time_delta::operator()(std::shared_ptr<dag_node_event> t0,
```
`const dag_node_event &` is sufficient

fknorr commented on src/runtime/omp/omp_queue.cpp:
```diff
@@ -69,8 +71,100 @@ bool is_contigous(id<3> offset, range<3> r, range<3> allocation_shape) {
+  friend class instrumentation_task_guard;
+
+  virtual profiler_clock::time_point get_time_point() const override {
+    return _time;
```
This can still be misused by calling `get_time_point()` before `wait()`, which will race on `_time`.

illuhad replied:
Not really - the whole point of the instrumentation API is that `instrumentation_set` guarantees that no instrumentation objects will be handed out until `wait()` has returned. This simplifies the design of instrumentations because it can be assumed that they will only be accessed (after initial setup) once they are ready.

fknorr replied:
Ok, then I'll be happy with an `assert(_signal.has_signalled())` in there.

fknorr commented on include/hipSYCL/runtime/dag_node.hpp:
```diff
@@ -93,6 +106,16 @@ class dag_node
+  /// Iterates across all operations that have actually been executed.
+  /// Precondition: is_submitted() returns true
+  template<class Handler>
+  void for_each_executed_operation(Handler h) {
```
This function is added but not currently used, is that intentional?

illuhad replied:
It provides a mechanism to access instrumented requirements. It's not yet used because it's unclear what should be returned for user-facing queries in case a requirement was optimized away, but I have some other PRs coming up that use it for internal runtime functionality.

fknorr reviewed (approved):
Aside from my comments above, this looks good to me!

illuhad commented:
@fknorr thanks, changing the `shared_ptr`s to const ref as you suggested is on my to do list and will happen before merge :)

sbalint98 committed:
Add atomic flag to signal channel

The change should make checking the signal channel more efficient on gcc version below 11.
see issue #601 for details

illuhad committed:
Merge pull request #602 from sbalint98/fix-signal-channel

Add atomic flag to signal channel

illuhad committed:
Change signature of timestamp delta instrumentation to use const ref

illuhad committed:
Add documentation w.r.t signal_channel performance issue

illuhad commented:
@hipSYCL-ci-bot test again please
